src/coreclr/jit/ee_il_dll.cpp

```diff
@@ -1615,45 +1615,40 @@ const char16_t* Compiler::eeGetShortClassName(CORINFO_CLASS_HANDLE clsHnd)
     return param.classNameWidePtr;
 }
 
-const WCHAR* Compiler::eeGetCPString(size_t strHandle)
+void Compiler::eePrintFrozenObjectDescription(const char* prefix, size_t handle)
 {
-#ifdef HOST_UNIX
-    return nullptr;
-#else
-    char buff[512 + sizeof(CORINFO_String)];
-
-    // make this bulletproof, so it works even if we are wrong.
-    if (ReadProcessMemory(GetCurrentProcess(), (void*)strHandle, buff, 4, nullptr) == 0)
+    const int maxStrSize = 64;
+    char      str[maxStrSize];
+    int       realLength = this->info.compCompHnd->objectToString((void*)handle, str, maxStrSize);
```

Comment:
The managed impl returns "bytes written" in . Should it be bytes written or real length?

Reply:
Ah, I guess it's a left-over from the initial impl that used to return real length (in utf16 chars) but then we switched to Utf8 for this API as it's only inteded to be used for debug/diagnostic purposes

Reply:
> it's only intended to be used for debug/diagnostic purposes so it should be bytesWritten now

Add a comment to corinfo.h?

It would be nice to follow a consistent naming and behavior pattern for these diagnostic-only methods that return diagnostic strings. We have `objectToString` and `appendClassName` that @jakobbotsch is touching in a separate PR.

Reply:
Personally, I very much dislike the "append" part of `appendClassName`. The caller can do this easily. If I had to pick I would probably pick:
1. The API always null terminates on its own
2. Lengths should be `size_t`
3. The API returns the actual length of a buffer necessary to hold the full result (including null terminator).
4. The API guarantees that `min(returnedSize, bufferSize)` characters were written to the buffer, including a null terminator.

I can move `appendClassName` to this model in my PR. I think we can name the APIs `printFrozenObject` and `printClassName`?

Any comments/suggestions @jkotas?

Reply:
I can also move `getMethodName` to this model in my PR. Today it has different behavior in checked and release builds, and JIT is actually only using it for printing helper function names.

Reply:
I think the SPMI recording could then be something like:
```cpp
Agnostic_AppendClassNameIn key;
key.Handle = classHandle;

size_t writtenSize = min(returnedSize, bufferSize);
int index = AppendClassName->GetIndex(key);
if (index != -1)
{
    Agnostic_AppendClassNameOut value = AppendClassName->GetItem(index);
    if (value.storedBufferLength >= bufferSize)
        return; // already storing more
}

Agnostic_AppendClassNameOut value;
value.returnedSize = returnedSize;
value.storedBufferSize = writtenSize;
value.bufferIndex = (DWORD)AppendClassName->AddBuffer(buffer, writtenSize);

if (index != -1)
    AppendClassName->Update(index, value);
else
    AppendClassName->Add(key, value);
```

that ensures we always store the largest call we saw.

Reply:
> Personally, I very much dislike the "append" part of appendClassName

I agree. The "append" part does not make sense.

> size_t printClassName(CORINFO_CLASS_HANDLE, char* buffer, size_t bufferSize)
> size_t printFrozenObject(void* handle, char* buffer, size_t bufferSize)

We should be consistent in whether we use "Object" or "FrozenObject" in the JIT-EE interface names. https://github.com/dotnet/runtime/pull/76112/files#diff-09491c1d7d4c7b20ad241d23d2b2ec182368b851cfb5f4ed4cdb5509dd6dd01fR291-R294 uses Object. I am not sure which one of the two options is better.

Reply:
@jakobbotsch can you explain the spmi part, what's wrong with the current code?

Reply:
I like `print` as the prefix to use for methods that are meant to return textual representation for logging or diagnostic purposes.

Reply:
I think it should work fine now that I look at it. Keying on the buffer size is somewhat odd, but hardly a problem given that we don't expect to see this JIT-EE function being used with many (or any) different buffer sizes.

The `appendClassName` version will need to be changed (it does not key on buffer size) so I'll try to factor them at that point.

Reply:
Also, we should be consistent in whether these methods zero-terminate the string when it does not fit or not. Either behavior is fine with me as long as it is consistent.